canal/sync.go

```diff
@@ -119,42 +111,26 @@ func (c *Canal) runSyncBinlog() error {
 				return errors.Trace(err)
 			}
 		case *replication.QueryEvent:
-			var (
-				mb    [][]byte
-				db    []byte
-				table []byte
-			)
-			regexps := []regexp.Regexp{*expCreateTable, *expAlterTable, *expRenameTable, *expDropTable, *expTruncateTable}
-			for _, reg := range regexps {
-				mb = reg.FindSubmatch(e.Query)
-				if len(mb) != 0 {
-					break
-				}
-			}
-			mbLen := len(mb)
-			if mbLen == 0 {
-				continue
-			}
-
-			// the first last is table name, the second last is database name(if exists)
-			if len(mb[mbLen-2]) == 0 {
-				db = e.Schema
-			} else {
-				db = mb[mbLen-2]
-			}
-			table = mb[mbLen-1]
-
-			savePos = true
-			force = true
-			c.ClearTableCache(db, table)
-			log.Infof("table structure changed, clear table cache: %s.%s\n", db, table)
-			if err = c.eventHandler.OnTableChanged(string(db), string(table)); err != nil && errors.Cause(err) != schema.ErrTableNotExist {
+			stmts, _, err := c.parser.Parse(string(e.Query), "", "")
+			if err != nil {
+				log.Errorf("parse query(%s) err %v", e.Query, err)
 				return errors.Trace(err)
 			}
-
-			// Now we only handle Table Changed DDL, maybe we will support more later.
-			if err = c.eventHandler.OnDDL(pos, e); err != nil {
-				return errors.Trace(err)
+			for _, stmt := range stmts {
+				nodes := parseStmt(stmt)
+				for _, node := range nodes {
+					if err = c.updateTable(node.db, node.table); err != nil {
+						return errors.Trace(err)
+					}
+				}
+				if len(nodes) > 0 {
+					savePos = true
```

Comment:
oh, should we also check `savePos` for `c.master.UpdateGTIDSet(e.GSet)` in L135~L136?
because we should not update GTID set for some query like `BEGIN`.